flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -219,6 +231,73 @@ public KubernetesWatch watchPodsAndDoCallback(
 				.watch(new KubernetesPodsWatcher(podCallbackHandler)));
 	}
 
+	@Override
+	public CompletableFuture<Void> createConfigMap(KubernetesConfigMap configMap) {
+		final String configMapName = configMap.getName();
+		return CompletableFuture.runAsync(
+			() -> this.internalClient.configMaps().inNamespace(namespace).create(configMap.getInternalResource()),
+			kubeClientExecutorService)
+			.exceptionally(throwable -> {
+				if (throwable != null) {
+					throw new CompletionException(
+						new KubernetesException("Failed to create ConfigMap " + configMapName, throwable));
+				}
+				return null;
+			});
+	}
+
+	@Override
+	public Optional<KubernetesConfigMap> getConfigMap(String name) {
+		final ConfigMap configMap = this.internalClient.configMaps().inNamespace(namespace).withName(name).get();
+		return configMap == null ? Optional.empty() : Optional.of(new KubernetesConfigMap(configMap));
+	}
+
+	@Override
+	public CompletableFuture<Boolean> checkAndUpdateConfigMap(
+			String configMapName,
+			FunctionWithException<KubernetesConfigMap, Optional<KubernetesConfigMap>, ?> function) {
```

Comment:
I am not yet fully convinced that `FunctionWithException` is the right choice. For the sake of completeness I repeat what I have written on an older discussion: I think we should only do `discardState` or something similar after we are sure that we have updated the `configMap` properly (e.g. removed the entry).

Reply:
Copy the comment above.

We do not need the FunctionWithException here. Only the update operation succeeded, we need to do the following external storage cleanup. For example, we could use kubeClient.checkAndUpdateConfigMap(configMapName, updateFunction).whenCompleteAsync() for this operation.